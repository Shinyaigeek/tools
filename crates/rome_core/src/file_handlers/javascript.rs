use super::{ExtensionHandler, Mime};
use std::fmt::Debug;

#[derive(Debug, PartialEq, Eq)]
pub struct JsFileHandler;

/// Features that belong to the language
#[derive(Debug)]
pub struct JsFileFeatures {
    /// Whether the file is executed as a script
    pub script: bool,
    /// Whether the file is executed as a module
    pub module: bool,
    /// Whether the TS grammar is supported
    pub typescript: bool,
    /// Whether the JSX syntax is supported
    pub jsx: bool,
    /// Whether the TSX syntax is supported
    pub tsx: bool,
}

impl ExtensionHandler for JsFileHandler {
    fn capabilities(&self) -> super::Capabilities {
        super::Capabilities {
            format: true,
            lint: false,
        }
    }

    fn language(&self) -> super::Language {
        super::Language::Js
    }

    fn mime(&self) -> super::Mime {
        Mime::Javascript
    }

    fn may_use_tabs(&self) -> bool {
        true
    }
}

impl Default for JsFileFeatures {
    fn default() -> Self {
        Self {
            module: true,
            script: false,
            jsx: false,
            tsx: false,
            typescript: false,
        }
    }
}

impl JsFileFeatures {
    /// Mark the file as module
    pub fn module(mut self) -> Self {
        self.module = true;
        self.script = false;
        self
    }

    /// Mark the file as a script
    ///
    /// With this, most of the other features are turned off
    pub fn script(mut self) -> Self {
        self.script = true;
        self.module = false;
        self.typescript = false;
        self.tsx = false;
        self.jsx = false;
        self
    }

    /// Enables TypeScript
    pub fn typescript(mut self) -> Self {
        self.script = false;
        self.module = true;
        self.typescript = true;
        self
    }
}
