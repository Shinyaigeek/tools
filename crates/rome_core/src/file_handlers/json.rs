use super::{ExtensionHandler, Mime};
#[derive(Debug, PartialEq, Eq)]
pub struct JsonFileHandler;

pub struct JsonFileFeatures {}

impl ExtensionHandler for JsonFileHandler {
    fn capabilities(&self) -> super::Capabilities {
        super::Capabilities {
            format: true,
            lint: false,
        }
    }

    fn language(&self) -> super::Language {
        super::Language::Json
    }

    fn mime(&self) -> super::Mime {
        Mime::Json
    }

    fn may_use_tabs(&self) -> bool {
        true
    }
}

impl Default for JsonFileFeatures {
    fn default() -> Self {
        Self {}
    }
}
