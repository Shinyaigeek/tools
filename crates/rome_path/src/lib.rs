//! This module is responsible to manage paths inside Rome.
//! It is a small wrapper around [path::PathBuf] but it is also able to
//! give additional information around the the file that holds:
//! - the [FileHandlers] for the specific file
//! - shortcuts to open/write to the file
use std::io::Read;
use std::{fs::File, io::Write, ops::Deref, path::PathBuf};

#[derive(Debug, Eq, Hash, PartialEq)]
pub struct RomePath {
    file: PathBuf,
}

impl Deref for RomePath {
    type Target = PathBuf;

    fn deref(&self) -> &Self::Target {
        &self.file
    }
}

impl RomePath {
    pub fn new(path_to_file: &str) -> Self {
        Self {
            file: PathBuf::from(path_to_file),
        }
    }

    // TODO: handle error with diagnostic?
    /// Opens a file and returns a [File] in write mode
    pub fn open(&self) -> File {
        File::open(&self.file).expect("cannot open the file to format")
    }

    /// Accepts a file opened in read mode and saves into it
    pub fn save(&mut self, content: &str) -> Result<(), std::io::Error> {
        let mut file_to_write = File::create(&self.file).unwrap();
        // TODO: handle error with diagnostic
        file_to_write.write_all(content.as_bytes())
    }

    /// Returns the contents of a file, if it exists
    pub fn get_buffer_from_file(&mut self) -> String {
        let mut file = self.open();
        let mut buffer = String::new();
        // we assume we have permissions
        file.read_to_string(&mut buffer)
            .expect("cannot read the file to format");

        buffer
    }
}
