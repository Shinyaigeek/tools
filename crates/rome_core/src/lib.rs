use crate::file_handlers::javascript::JsFileFeatures;
use crate::file_handlers::json::JsonFileFeatures;
use crate::file_handlers::unknown::UnknownFileHandler;
use crate::file_handlers::{javascript::JsFileHandler, ExtensionHandler, Language};
use file_handlers::json::JsonFileHandler;
use rome_path::RomePath;
use std::collections::HashMap;

pub mod file_handlers;

/// Features available for each language
struct Features {
    js: JsFileHandler,
    json: JsonFileHandler,
    unknown: UnknownFileHandler,
}

pub struct App {
    /// features available throughout the application
    features: Features,
    /// Associate a JavaScript file with its features
    js_files: HashMap<RomePath, JsFileFeatures>,
    /// Associate a JSON file with its features
    json_files: HashMap<RomePath, JsonFileFeatures>,
}

impl Default for App {
    fn default() -> Self {
        Self {
            features: Features {
                js: JsFileHandler {},
                json: JsonFileHandler {},
                unknown: UnknownFileHandler::default(),
            },
            js_files: HashMap::new(),
            json_files: HashMap::new(),
        }
    }
}

impl App {
    pub fn new() -> Self {
        Default::default()
    }

    /// Store a JavaScript file with its features
    pub fn store_js_file(&mut self, path_to_file: &str, file_features: JsFileFeatures) {
        let path = RomePath::new(path_to_file);
        self.js_files.insert(path, file_features);
    }

    /// Return the [JsFileFeatures] that belong to the given file
    pub fn get_js_file(&self, path: &RomePath) -> Option<&JsFileFeatures> {
        self.js_files.get(path)
    }

    /// Store a JSON file with its features
    pub fn store_json_file(&mut self, path_to_file: &str) {
        let path = RomePath::new(path_to_file);
        let features = JsonFileFeatures::default();
        self.json_files.insert(path, features);
    }

    /// Return the [JsonFileFeatures] that belong to the given file
    pub fn get_json_file(&self, path: &RomePath) -> Option<&JsonFileFeatures> {
        self.json_files.get(path)
    }

    /// Return a [Language] from a string
    pub fn get_language<L: Into<Language>>(&self, file_extension: L) -> Language {
        file_extension.into()
    }

    /// Check if the current language is supported
    pub fn is_language_supported<L: Into<Language>>(&self, file_extension: L) -> bool {
        Language::Unknown != file_extension.into()
    }

    /// Return the features that are available for JavaScript
    pub fn get_js_features(&self) -> &JsFileHandler {
        &self.features.js
    }

    /// Return the features that are available for JSON
    pub fn get_json_features(&self) -> &JsonFileHandler {
        &self.features.json
    }

    /// Features available to a language that is not supported
    pub fn get_unknown_features(&self) -> &UnknownFileHandler {
        &self.features.unknown
    }
}
