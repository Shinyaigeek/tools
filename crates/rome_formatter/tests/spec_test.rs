use rome_core::file_handlers::Language;
use rome_core::App;
use rome_formatter::{format_js_file, format_json_file, FormatError, FormatOptions};
use rome_path::RomePath;
use std::fs;
use std::path::Path;

/// [insta.rs](https://insta.rs/docs) snapshot testing
///
/// For better development workflow, run
/// `cargo watch -i '*.new' -x 'test -p rome_formatter formatter'`
///
/// To review and commit the snapshots, `cargo install cargo-insta`, and run
/// `cargo insta review` or `cargo insta accept`
///
/// The input and the expected output are stored as dedicated files in the `tests/specs` directory where
/// the input file name is `{spec_name}.json` and the output file name is `{spec_name}.json.snap`.
///
/// Specs can be grouped in directories by specifying the directory name in the spec name. Examples:
///
/// # Examples
///
/// * `json/null` -> input: `tests/specs/json/null.json`, expected output: `tests/specs/json/null.json.snap`
/// * `null` -> input: `tests/specs/null.json`, expected output: `tests/specs/null.json.snap`
pub fn run(spec_input_file: &str, _: &str, file_type: &str) {
    let mut app = App::new();
    let file_path = &spec_input_file;
    let spec_input_file = Path::new(spec_input_file);

    assert!(
        spec_input_file.is_file(),
        "The input '{}' must exist and be a file.",
        spec_input_file.display()
    );

    let mut path = RomePath::new(file_path);
    let language = app.get_language(path.extension().unwrap());
    let formatted_result = match language {
        Language::Js | Language::Ts => {
            let module = file_type == "module";

            app.store_js_file(file_path, module);

            format_js_file(&mut path, FormatOptions::default(), &app)
        }
        Language::Json => {
            app.store_json_file(file_path);

            format_json_file(&mut path, FormatOptions::default(), &app)
        }
        Language::Unknown => Err(FormatError::UnsupportedLanguage),
    };
    let file_name = spec_input_file.file_name().unwrap().to_str().unwrap();
    let input = fs::read_to_string(file_path).unwrap();
    // we ignore the error for now
    let snapshot = format!(
        "# Input\n{}\n---\n# Output\n{}",
        input,
        formatted_result.unwrap().code()
    );

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => spec_input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });
}
