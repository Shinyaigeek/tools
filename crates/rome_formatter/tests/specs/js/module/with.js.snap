---
source: crates/rome_formatter/tests/spec_test.rs
assertion_line: 68
expression: with.js

---
# Input
with (   b)

{
	5
}

---
# Output
with (   b)

{
	5
}

